File: src/lib/deck.ts
import type { RNG } from './rng'

export type Suit = 'clubs' | 'diamonds' | 'hearts' | 'spades'
export type Rank = '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | 'J' | 'Q' | 'K' | 'A'

export type Card = {
  rank: Rank
  suit: Suit
  faceDown: boolean
}

export type Shoe = {
  cards: Card[]
  numDecks: number
  dealtCount: number
}

const SUITS: Suit[] = ['clubs', 'diamonds', 'hearts', 'spades']
const RANKS: Rank[] = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K', 'A']

export function makeCard(rank: Rank, suit: Suit, faceDown = false): Card {
  return { rank, suit, faceDown }
}

export function makeDeck(): Card[] {
  const deck: Card[] = []
  for (const suit of SUITS) {
    for (const rank of RANKS) {
      deck.push(makeCard(rank, suit))
    }
  }
  return deck
}

export function shuffle(cards: Card[], rng: RNG): Card[] {
  const result = [...cards]
  for (let i = result.length - 1; i > 0; i--) {
    const j = Math.floor(rng.next() * (i + 1))
    const tmp = result[i]!
    result[i] = result[j]!
    result[j] = tmp
  }
  return result
}

export function makeShoe(numDecks: number, rng: RNG): Shoe {
  const cards: Card[] = []
  for (let d = 0; d < numDecks; d++) {
    cards.push(...makeDeck())
  }
  return {
    cards: shuffle(cards, rng),
    numDecks,
    dealtCount: 0,
  }
}

export function dealCard(shoe: Shoe, faceDown = false): { card: Card; shoe: Shoe } {
  if (shoe.dealtCount >= shoe.cards.length) {
    throw new Error('Shoe is empty')
  }
  const card = { ...shoe.cards[shoe.dealtCount]!, faceDown }
  return {
    card,
    shoe: { ...shoe, dealtCount: shoe.dealtCount + 1 },
  }
}

export function cardsRemaining(shoe: Shoe): number {
  return shoe.cards.length - shoe.dealtCount
}

export function penetration(shoe: Shoe): number {
  return shoe.dealtCount / shoe.cards.length
}

File: games/blackjack/logic/Deck.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace TableGames.Blackjack.Logic;

/// <summary>
/// Six-deck shoe. Fisher-Yates shuffle. Tracks penetration for reshuffle trigger.
/// Spec: game-spec.md Section 2.
/// </summary>
public class Deck
{
    private const int NumDecks = 6;
    private const int CardsPerDeck = 52;
    private const int TotalCards = NumDecks * CardsPerDeck; // 312
    private const int ShuffleThreshold = TotalCards / 2;    // 156

    private Card[] _shoe;
    private int _topIndex; // next card to deal is at _shoe[_topIndex]

    public int TotalDealt { get; private set; }
    public bool ShufflePending => TotalDealt >= ShuffleThreshold;
    public int CardsRemaining => TotalCards - _topIndex;

    public Deck()
    {
        _shoe = BuildSortedShoe();
        TotalDealt = 0;
        _topIndex = 0;
        Shuffle();
    }

    /// <summary>Peek all remaining cards without dealing them. For testing only.</summary>
    public Card[] PeekAllCards()
    {
        var result = new Card[CardsRemaining];
        Array.Copy(_shoe, _topIndex, result, 0, CardsRemaining);
        return result;
    }

    /// <summary>Deal the top card from the shoe.</summary>
    public Card Deal()
    {
        if (_topIndex >= TotalCards)
            throw new InvalidOperationException("Shoe is empty — this is a fatal error state per spec section 2.4");
        var card = _shoe[_topIndex++];
        TotalDealt++;
        return card;
    }

    /// <summary>
    /// Fisher-Yates in-place shuffle of all 312 cards.
    /// Resets TotalDealt to 0 and ShufflePending to false.
    /// Spec: game-spec.md Section 2.5.
    /// </summary>
    public void Shuffle()
    {
        // Rebuild full shoe from index 0
        _shoe = BuildSortedShoe();
        _topIndex = 0;
        TotalDealt = 0;

        // Fisher-Yates
        var rng = new Random();
        for (int i = TotalCards - 1; i > 0; i--)
        {
            int j = rng.Next(i + 1);
            (_shoe[i], _shoe[j]) = (_shoe[j], _shoe[i]);
        }
    }

    private static Card[] BuildSortedShoe()
    {
        var shoe = new Card[TotalCards];
        int index = 0;
        for (int deck = 0; deck < NumDecks; deck++)
        {
            foreach (Suit suit in Enum.GetValues<Suit>())
            {
                for (int rank = 1; rank <= 13; rank++)
                {
                    shoe[index++] = new Card(rank, suit);
                }
            }
        }
        return shoe;
    }
}
